### ISSUE TITLE
Add HonoClient#disconnect() method

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#500]
Currently there is a `HonoClient#connect()` method, but no `HonoClient#disconnect()` method.

A "disconnect" method is for example needed in a scenario where the open frame of the connection contained permissions from an authorization service and these permissions in the service have changed since opening the connection, making it necessary to disconnect and re-open the connection.

I could imagine a simple method with no parameters and return value here:
``
void disconnect();
``
Notification of a finished disconnect will be done on the `disconnectHandler` supplied to the `connect` method.

**Comment by sophokles73** [comment]
What would/should be the difference to invoking `HonoClient.shutdown()`?

**Comment by calohmn** [comment]
A `HonoClient.shutdown()` invocation causes all subsequent `HonoClient.connect` invocations to fail. I.e. the HonoClient can't be used anymore after that.

A `disconnect` should allow further `connect` invocations.

**Comment by sophokles73** [comment]
should be fixed in 1550e8950be4f695cce929584fd35361fe0b039c

@calohmn can you verify and then close this issue, if it works for you?

**Comment by calohmn** [comment]
Re-test successful.

What is perhaps somewhat unexpected, is that the `disconnect completionHandler` is called directly after having sent the "close" frame, without waiting for a "close" frame response or for the underlying transport connection to be disconnected.

Log output where the `disconnect completionHandler` is doing a new `connect` - see the late `closed connection to server` log line:
````
16:08:25.569 INFO  o.e.hono.example.ExampleReceiver - ##### call disconnect
16:08:25.574 INFO  o.e.hono.client.impl.HonoClientImpl - closing connection to server [192.168.99.100:30071]...

  16:08:25.576 INFO  o.e.hono.example.ExampleReceiver - ##### disconnect completionHandler succeeded; now reconnect...

16:08:25.582 DEBUG o.e.hono.client.impl.HonoClientImpl - starting attempt [#1] to connect to server [192.168.99.100:30071]
16:08:25.582 DEBUG o.e.h.config.ClientConfigProperties - using certificates from file [[..]trusted-certs.pem] as trust anchor
16:08:25.583 DEBUG o.e.h.c.impl.ConnectionFactoryImpl - enabling secure protocol [TLSv1.2]
16:08:25.583 DEBUG o.e.h.c.impl.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://192.168.99.100:30071]

  16:08:25.590 INFO  o.e.hono.client.impl.HonoClientImpl - closed connection to server [192.168.99.100:30071]

16:08:25.944 DEBUG o.e.h.c.impl.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://192.168.99.100:30071], opening connection ...
16:08:25.948 DEBUG o.e.h.c.impl.ConnectionFactoryImpl - connection to container [Router.qdrouterd-756f57946-k48tp] at [amqps://192.168.99.100:30071] open
16:08:25.957 DEBUG o.e.h.client.impl.AbstractHonoClient - receiver open [source: telemetry/PERF_TEST_TENANT]
````

But IMHO that shouldn't be a problem and will speed up reconnect operations a bit.

**Comment by sophokles73** [comment]
> What is perhaps somewhat unexpected, is that the disconnect completionHandler is called directly after having sent the "close" frame, without waiting for a "close" frame response or for the underlying transport connection to be disconnected.

this is because we cannot be sure that the peer will actually send one and we actually do not care as well because from our point of view, the connection is closed anyway ...
